src/api/cpp/cvc5.h

```diff
@@ -4372,6 +4547,36 @@ class CVC5_EXPORT Solver
    */
   Statistics getStatistics() const;
 
+  /* .................................................................... */
+  /* Optimization                                                         */
+  /* .................................................................... */
+
+  /**
+   * Run optimization for the objectives
+   * @param objCombination for multi-objective optimization,
+   *  indicate whether it's a BOX optimization,
+   *  a LEXICOGRAPHIC optimization or a PARETO optimization
+   * @return a pair <Result, vector<OptimizationResult>>,
+   *  where Result indicates the overall optimization result,
+   *  for BOX, if any of the objective is UNSAT or UNKNOWN,
+   *    it's UNSAT/UNKNOWN,
+   *  for LEXICOGRAPHIC/PARETO, if the assertions are UNSAT/UNKNOWN,
+   *    it's UNSAT/UNKNOWN,
+   *  vector<OptimizationResult> stores the results of individual objectives
+   */
+  std::pair<Result, std::vector<OptimizationResult>> checkOpt(
+      const std::vector<std::pair<Term, ObjectiveType>>& objectives,
+      ObjectiveCombination objCombination = LEXICOGRAPHIC) const;
+
+  /**
+   * Retrive the next Pareto solution,
+   * valid only after checkOpt is called with PARETO and SAT is returned.
+   * @return a pair <Result, vector<OptimizationResult>>,
+   *  where Result indicates the overall optimization result,
+   *  if the possible solutions are exhausted, it's UNSAT.
+   */
+  std::pair<Result, std::vector<OptimizationResult>> nextParetoSolution() const;
```

Comment:
Sure, but see the above comment 👆

Reply:
I don't follow. Can you explain?

Reply:
the comment in the post above.

Reply:
I mean this PR should be closed as I think we should figure out APIs only after the online optimization is implemented, instead of just arguing on the air.

Reply:
Ok, makes sense.

Reply:
We should discuss this in more detail.

In cvc5, we will need to extend the "SMT mode": https://github.com/cvc5/cvc5/blob/master/src/smt/smt_mode.h
This tracks the abstract state we are in on the diagram Figure 4.1 in page 52 of https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf.
Notice that we have already informally extended this state diagram with two more modes "ABDUCT" and "INTERPOL".  The change of modes is managed at the SmtEngine level, e.g. see the comment here: https://github.com/cvc5/cvc5/blob/master/src/smt/smt_engine.cpp#L1719.

For optimization, we will likely need one new mode "SAT_OPTIMIZED".  An optimize-sat that returns "sat" goes to this mode, an "unsat" goes to the normal UNSAT mode.  The tricky part is that models *should be* available in "OPTIMIZED_SAT" mode; solutions for the last optimization problem are *only* available in "SAT_OPTIMIZED" mode.

Reply:
As discussed in the meeting, we should not have the semantics of commands depend on options. This means that `check-sat` always means "check satisfibility without optimization" and `optimize-sat` means "check for satisfiability with optimization.  Moreover we provide the kind of optimization as an argument to `optimize-sat`.

Reply:
but this does not incorporate well with (push) and (pop), and since the whole solver is in optimization mode we need to suppose this as well. as for the (check-sat), we can forbid this in optimization mode and force the user to use (optimize-sat). But supplying targets as arguments will definitely kill possibilities in (push) & (pop).

Reply:
Oh I mean if we use (optimize-sat (target 1) (target 2)...), then it's hard to re-use the clauses next-time. 
of course we can implement an internal caching mechanism, but in this case every time the use does the pop we need to hook that and clear the corresponding cache.

Reply:
Like with ordinary models, push and pop destroy the state of the current model, as it is no longer valid in the current context.  It also will destroy the state of the last optimization query; the objective has to be supplied again to another `optimize-sat` command.  We do not want to do anything fancier than this. 

Again, my recommendation is for us to focus on implementing the basic optimization methods without considering getting multiple solutions.

Reply:
In particular, the important detail to notice is the edges marked "p" for push/pop in Figure 4.1 of https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf.  This means that for:
```
(check-sat) ; returns "sat"
(pop)
...  the model is *not* available here.
```

Reply:
sure, but the fact is that we have multiple objectives working, but the single query 😂 (we only have one offline BV binary search optimizer, but this is not enough, we should implement an online optimization for LIRA). 

for multiple objectives, it would be good to incorporate with push and pop, but if we don't want to it's ok (it's beneficial to lexico and box optimizations as we can use some of the learnt clauses, but it's hard to incorporate Pareto with push/pop).

Reply:
> In particular, the important detail to notice is the edges marked "p" for push/pop in Figure 4.1 of https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf. This means that for:
> 
> ```
> (check-sat) ; returns "sat"
> (pop)
> ...  the model is *not* available here.
> ```

Right right right, we need to check-sat again, but this time we used some of the learnt clauses in this level so it's faster. I think this should be incorporated into optimization as well so that we don't do everything again.